---

_You are currently looking at **version 1.2** of this notebook. To download notebooks and datafiles, as well as get help on Jupyter notebooks in the Coursera platform, visit the [Jupyter Notebook FAQ](https://www.coursera.org/learn/python-social-network-analysis/resources/yPcBs) course resource._

---

# Assignment 4

In [1]:
import networkx as nx
import pandas as pd
import numpy as np
import pickle
import operator
import warnings
warnings.filterwarnings('ignore')

---

## Part 1 - Random Graph Identification

For the first part of this assignment you will analyze randomly generated graphs and determine which algorithm created them.

In [2]:
P1_Graphs = pickle.load(open('A4_graphs','rb'))
P1_Graphs

<br>
`P1_Graphs` is a list containing 5 networkx graphs. Each of these graphs were generated by one of three possible algorithms:
* Preferential Attachment (`'PA'`)
* Small World with low probability of rewiring (`'SW_L'`)
* Small World with high probability of rewiring (`'SW_H'`)

Anaylze each of the 5 graphs and determine which of the three algorithms generated the graph.

*The `graph_identification` function should return a list of length 5 where each element in the list is either `'PA'`, `'SW_L'`, or `'SW_H'`.*

In [3]:
# PA - The nodes are attached to the node which has maximum neighbours. 
# This will result in the degree of some nodes to be much more higher than other
# SW_L prob - Avg._short_path_length and Avg._Clus_coeff, both will be high.
# SW_H prob - Avg._short_path_length and Avg._Clus_coeff, both will be low.

In [4]:
G_PA=[]
top5_PA=[]
G_Deg=[]
G_ACL=[]
G_ASL=[]
for i,G in enumerate(P1_Graphs):
    G_PA.append(sorted(nx.preferential_attachment(G), key=operator.itemgetter(2), reverse=True))
    top5_PA.append([PA[2] for PA in G_PA[i][:5]])
    G_Deg.append(sorted(G.degree().items(), key=operator.itemgetter(1), reverse=True)[0][1])
    G_ACL.append(nx.average_clustering(G))
    G_ASL.append(nx.average_shortest_path_length(G))

In [5]:
# G_Deg=[]
# for i,G in enumerate(P1_Graphs):
#     if i == 0:
#         G_Deg.append(sorted(G.degree().items(), key=operator.itemgetter(1), reverse=True)[0][1])

In [6]:
# top5_PA=[]
# for i in range(5):
#     top5_PA.append([PA[2] for PA in G_PA[i][:5]])   # sorted_G_PA

In [7]:
top5_PA

[[4900, 3000, 2900, 2700, 2600],
 [156, 156, 156, 156, 156],
 [36, 36, 36, 36, 36],
 [4692, 4324, 4232, 4048, 3864],
 [99, 99, 88, 88, 88]]

In [8]:
G_Deg

[100, 13, 6, 92, 11]

In [9]:
G_ACL

[0.03167539146454044,
 0.5642419635919628,
 0.4018222222222227,
 0.03780379975223251,
 0.0033037037037037037]

In [10]:
G_ASL

[4.099161161161161,
 5.089871871871872,
 9.378702269692925,
 3.1048046283934134,
 5.0785509568313305]

In [11]:
list(zip(G_Deg,G_ACL,G_ASL))

[(100, 0.03167539146454044, 4.099161161161161),
 (13, 0.5642419635919628, 5.089871871871872),
 (6, 0.4018222222222227, 9.378702269692925),
 (92, 0.03780379975223251, 3.1048046283934134),
 (11, 0.0033037037037037037, 5.0785509568313305)]

In [12]:
['PA', 'SW_L', 'SW_L', 'PA', 'SW_H']

['PA', 'SW_L', 'SW_L', 'PA', 'SW_H']

In [13]:
def graph_identification():
    
    # Your Code Here    
    return ['PA', 'SW_L', 'SW_L', 'PA', 'SW_H']
    # Your Answer Here

---

## Part 2 - Company Emails

For the second part of this assignment you will be workking with a company's email network where each node corresponds to a person at the company, and each edge indicates that at least one email has been sent between two people.

The network also contains the node attributes `Department` and `ManagementSalary`.

`Department` indicates the department in the company which the person belongs to, and `ManagementSalary` indicates whether that person is receiving a management position salary.

In [14]:
G = nx.read_gpickle('email_prediction.txt')

print(nx.info(G))

Name: 
Type: Graph
Number of nodes: 1005
Number of edges: 16706
Average degree:  33.2458


In [15]:
G.nodes(data=True)

[(0, {'Department': 1, 'ManagementSalary': 0.0}),
 (1, {'Department': 1, 'ManagementSalary': nan}),
 (2, {'Department': 21, 'ManagementSalary': nan}),
 (3, {'Department': 21, 'ManagementSalary': 1.0}),
 (4, {'Department': 21, 'ManagementSalary': 1.0}),
 (5, {'Department': 25, 'ManagementSalary': nan}),
 (6, {'Department': 25, 'ManagementSalary': 1.0}),
 (7, {'Department': 14, 'ManagementSalary': 0.0}),
 (8, {'Department': 14, 'ManagementSalary': nan}),
 (9, {'Department': 14, 'ManagementSalary': 0.0}),
 (10, {'Department': 9, 'ManagementSalary': 0.0}),
 (11, {'Department': 14, 'ManagementSalary': 0.0}),
 (12, {'Department': 14, 'ManagementSalary': 1.0}),
 (13, {'Department': 26, 'ManagementSalary': 1.0}),
 (14, {'Department': 4, 'ManagementSalary': nan}),
 (15, {'Department': 17, 'ManagementSalary': 0.0}),
 (16, {'Department': 34, 'ManagementSalary': 0.0}),
 (17, {'Department': 1, 'ManagementSalary': 0.0}),
 (18, {'Department': 1, 'ManagementSalary': nan}),
 (19, {'Department': 14, 'Ma

In [16]:
len(G.edges())
# edges do not have any attributes

16706

In [17]:
df_1 = pd.DataFrame()

In [18]:
df_1['Node'] = G.nodes()
df_1['Department'] = pd.Series(G.nodes(data=True)).map(lambda x: x[1]['Department'])
# df_1['Department'] = pd.Series(nx.get_node_attributes(G, 'Department'))
df_1['ManagementSalary'] = pd.Series(G.nodes(data=True)).map(lambda x: x[1]['ManagementSalary'])
# df_1['ManagementSalary'] = pd.Series(nx.get_node_attributes(G, 'ManagementSalary'))

df_1.head()

Node  Department  ManagementSalary
0     0           1               0.0
1     1           1               NaN
2     2          21               NaN
3     3          21               1.0
4     4          21               1.0

In [19]:
df_1['clustering'] = pd.Series(nx.clustering(G))
df_1['degree'] = pd.Series(G.degree())

df_1.head()

Node  Department  ManagementSalary  clustering  degree
0     0           1               0.0    0.276423      44
1     1           1               NaN    0.265306      52
2     2          21               NaN    0.297803      95
3     3          21               1.0    0.384910      71
4     4          21               1.0    0.318691      96

In [20]:
df_train = df_1[df_1['ManagementSalary'].notnull()]

df_train.head()

Node  Department  ManagementSalary  clustering  degree
0     0           1               0.0    0.276423      44
3     3          21               1.0    0.384910      71
4     4          21               1.0    0.318691      96
6     6          25               1.0    0.155183     115
7     7          14               0.0    0.287785      72

In [21]:
df_test = df_1[df_1['ManagementSalary'].isnull()]

df_test.head()

Node  Department  ManagementSalary  clustering  degree
1      1           1               NaN    0.265306      52
2      2          21               NaN    0.297803      95
5      5          25               NaN    0.107002     171
8      8          14               NaN    0.447059      37
14    14           4               NaN    0.215784      80

In [22]:
df_train.shape[0], df_test.shape[0], df_train.shape[0]+df_test.shape[0]

(753, 252, 1005)

In [23]:
import sklearn
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import train_test_split

#  use predict_proba() from logisticregression to get the probability prediction

In [24]:
X_train, X_test, y_train, y_test = (train_test_split(df_train[['Department', 'clustering', 'degree']], 
                                                     df_train['ManagementSalary'], random_state=23))

In [25]:
model = LogisticRegression(random_state=24).fit(X_train, y_train)

y_pred = model.predict(X_test)
y_pred_prob = model.predict_proba(X_test)

print('Train Accuracy_score: {}'.format(model.score(X_train, y_train)))
print('Test Accuracy_score: {}'.format(model.score(X_test, y_test)))
print('ROC_AUC_score: {}'.format(roc_auc_score(y_test,y_pred_prob[:,-1])))


Train Accuracy_score: 0.8953900709219859
Test Accuracy_score: 0.8571428571428571
ROC_AUC_score: 0.8574928542262147


In [26]:
RFC = RandomForestClassifier(n_estimators=400, random_state=24).fit(X_train, y_train)

y_pred = RFC.predict(X_test)
y_pred_prob = RFC.predict_proba(X_test)

print('Train Accuracy_score: {}'.format(RFC.score(X_train, y_train)))
print('Test Accuracy_score: {}'.format(RFC.score(X_test, y_test)))
print('ROC_AUC_score: {}'.format(roc_auc_score(y_test,y_pred_prob[:,-1])))


Train Accuracy_score: 1.0
Test Accuracy_score: 0.8571428571428571
ROC_AUC_score: 0.8912821559820335


In [27]:
X_sub = df_test[['Department', 'clustering', 'degree']]

df_test['Prob_0'] = RFC.predict_proba(X_sub)[:,0]
df_test['Prob_1'] = RFC.predict_proba(X_sub)[:,1]

In [28]:
RFC.classes_

array([ 0.,  1.])

In [29]:
df_test.head()

Node  Department  ManagementSalary  clustering  degree  Prob_0  Prob_1
1      1           1               NaN    0.265306      52  0.9750  0.0250
2      2          21               NaN    0.297803      95  0.0750  0.9250
5      5          25               NaN    0.107002     171  0.0175  0.9825
8      8          14               NaN    0.447059      37  0.8975  0.1025
14    14           4               NaN    0.215784      80  0.9100  0.0900

In [30]:
df_test.shape[0]

252

In [31]:
df_test['Prob_1'] = df_test['Prob_1'].map(lambda x: round(x,1))
df_test['Prob_0'] = df_test['Prob_0'].map(lambda x: round(x,1))

In [32]:
df_test.head()

Node  Department  ManagementSalary  clustering  degree  Prob_0  Prob_1
1      1           1               NaN    0.265306      52     1.0     0.0
2      2          21               NaN    0.297803      95     0.1     0.9
5      5          25               NaN    0.107002     171     0.0     1.0
8      8          14               NaN    0.447059      37     0.9     0.1
14    14           4               NaN    0.215784      80     0.9     0.1

In [33]:
round(0.0250,1), round(0.9825,1)

(0.0, 1.0)

In [34]:
type(df_test['Prob_1'])

pandas.core.series.Series

In [35]:
df_test['Prob_1'][:5]

1     0.0
2     0.9
5     1.0
8     0.1
14    0.1
Name: Prob_1, dtype: float64

### Part 2A - Salary Prediction

Using network `G`, identify the people in the network with missing values for the node attribute `ManagementSalary` and predict whether or not these individuals are receiving a management position salary.

To accomplish this, you will need to create a matrix of node features using networkx, train a sklearn classifier on nodes that have `ManagementSalary` data, and predict a probability of the node receiving a management salary for nodes where `ManagementSalary` is missing.



Your predictions will need to be given as the probability that the corresponding employee is receiving a management position salary.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.88 or higher will receive full points, and with an AUC of 0.82 or higher will pass (get 80% of the full points).

Using your trained classifier, return a series of length 252 with the data being the probability of receiving management salary, and the index being the node id.

    Example:
    
        1       1.0
        2       0.0
        5       0.8
        8       1.0
            ...
        996     0.7
        1000    0.5
        1001    0.0
        Length: 252, dtype: float64

In [36]:
def salary_predictions():
    
    # Your Code Here
    import sklearn
    from sklearn.linear_model import LogisticRegression
    from sklearn.ensemble import RandomForestClassifier
    from sklearn.metrics import roc_auc_score
    from sklearn.model_selection import train_test_split
    
    df_1 = pd.DataFrame()
    df_1['Node'] = G.nodes()
    df_1['Department'] = pd.Series(G.nodes(data=True)).map(lambda x: x[1]['Department'])
    df_1['ManagementSalary'] = pd.Series(G.nodes(data=True)).map(lambda x: x[1]['ManagementSalary'])
    df_1['clustering'] = pd.Series(nx.clustering(G))
    df_1['degree'] = pd.Series(G.degree())
    df_train = df_1[df_1['ManagementSalary'].notnull()]
    df_test = df_1[df_1['ManagementSalary'].isnull()]
        
    X_train, X_test, y_train, y_test = (train_test_split(df_train[['Department', 'clustering', 'degree']], 
                                                     df_train['ManagementSalary'], random_state=23))
    RFC = RandomForestClassifier(n_estimators=400, random_state=24).fit(X_train, y_train)

    X_sub = df_test[['Department', 'clustering', 'degree']]

    df_test['Prob_0'] = RFC.predict_proba(X_sub)[:,0]
    df_test['Prob_1'] = RFC.predict_proba(X_sub)[:,1]
    df_test['Prob_1'] = df_test['Prob_1'].map(lambda x: round(x,1))
    
    return df_test['Prob_1']# Your Answer Here

In [37]:
salary_predictions()[:5]

1     0.0
2     0.9
5     1.0
8     0.1
14    0.1
Name: Prob_1, dtype: float64

### Part 2B - New Connections Prediction

For the last part of this assignment, you will predict future connections between employees of the network. The future connections information has been loaded into the variable `future_connections`. The index is a tuple indicating a pair of nodes that currently do not have a connection, and the `Future Connection` column indicates if an edge between those two nodes will exist in the future, where a value of 1.0 indicates a future connection.

In [38]:
future_connections = pd.read_csv('Future_Connections.csv', index_col=0, converters={0: eval})
future_connections.head(10)

Future Connection
(6, 840)                  0.0
(4, 197)                  0.0
(620, 979)                0.0
(519, 872)                0.0
(382, 423)                0.0
(97, 226)                 1.0
(349, 905)                0.0
(429, 860)                0.0
(309, 989)                0.0
(468, 880)                0.0

In [39]:
future_connections.shape

(488446, 1)

In [40]:
future_connections.isnull().sum()

Future Connection    122112
dtype: int64

In [41]:
future_connections[future_connections.notnull()].head()

Future Connection
(6, 840)                  0.0
(4, 197)                  0.0
(620, 979)                0.0
(519, 872)                0.0
(382, 423)                0.0

In [42]:
future_connections[future_connections['Future Connection']==1].head()

Future Connection
(97, 226)                 1.0
(342, 473)                1.0
(206, 303)                1.0
(590, 725)                1.0
(170, 430)                1.0

In [43]:
future_connections[future_connections.isnull()].head()

Future Connection
(6, 840)                  NaN
(4, 197)                  NaN
(620, 979)                NaN
(519, 872)                NaN
(382, 423)                NaN

In [44]:
# G.edges()

In [45]:
df_edge = pd.DataFrame(index=G.edges())

In [46]:
df_edge.head()

Empty DataFrame
Columns: []
Index: [(0, 1), (0, 17), (0, 316), (0, 146), (0, 581)]

In [47]:
df_index = df_edge.index.tolist()

In [48]:
FC_index = future_connections.index.tolist()

In [49]:
set(df_index).intersection(FC_index)

set()

In [50]:
set(df_index) & set(FC_index)

set()

In [51]:
# common_index=[i for i in FC_index if i in df_index]
# commputationaly expensive process

In [52]:
# future_connections[future_connections.index == df_edge.index]['Future Connection']

In [53]:
df_edge.index

Index([     (0, 1),     (0, 17),    (0, 316),    (0, 146),    (0, 581),
          (0, 268),    (0, 221),    (0, 218),     (0, 18),    (0, 734),
       ...
        (971, 971),  (976, 981),  (976, 976),  (977, 977),  (979, 979),
        (980, 980),  (981, 981),  (989, 989), (990, 1001),  (992, 992)],
      dtype='object', length=16706)

In [54]:
for (u,v) in df_edge.index:
    df_edge.loc[(u,v),'#common_neighbors'] = len(list(nx.common_neighbors(G, u, v)))

In [55]:
df_edge['preferential attachment'] = [i[2] for i in nx.preferential_attachment(G, df_edge.index)]
df_edge['resource_allocation_index'] = [i[2] for i in nx.resource_allocation_index(G, df_edge.index)]
df_edge['jaccard_coefficient'] = [i[2] for i in nx.jaccard_coefficient(G, df_edge.index)]

In [56]:
# df_edge['adamic_adar_index'] = [i[2] for i in nx.adamic_adar_index(G, df_edge.index)]
# Can't us this feature, getting ZeroDivisionError: float division by zero

In [57]:
df_edge.head()

#common_neighbors  preferential attachment  \
(0, 1)                 14.0                     2288   
(0, 17)                21.0                     4884   
(0, 316)               17.0                     2068   
(0, 146)                9.0                     1320   
(0, 581)                3.0                     1408   

          resource_allocation_index  jaccard_coefficient  
(0, 1)                     0.246179             0.179487  
(0, 17)                    0.486685             0.161538  
(0, 316)                   0.326851             0.242857  
(0, 146)                   0.181323             0.147541  
(0, 581)                   0.051320             0.043478

In [58]:
df_edge.isnull().sum()

#common_neighbors            0
preferential attachment      0
resource_allocation_index    0
jaccard_coefficient          0
dtype: int64

In [59]:
df_edge.shape

(16706, 4)

In [60]:
df_fc = pd.DataFrame(index=future_connections.index)

In [ ]:
df_fc.head()

Empty DataFrame
Columns: []
Index: [(6, 840), (4, 197), (620, 979), (519, 872), (382, 423)]

In [ ]:
for (u,v) in df_fc.index:
    df_fc.loc[(u,v),'#common_neighbors'] = len(list(nx.common_neighbors(G, u, v)))

In [ ]:
df_fc['preferential attachment'] = [i[2] for i in nx.preferential_attachment(G, df_fc.index)]
df_fc['resource_allocation_index'] = [i[2] for i in nx.resource_allocation_index(G, df_fc.index)]
df_fc['jaccard_coefficient'] = [i[2] for i in nx.jaccard_coefficient(G, df_fc.index)]

In [ ]:
df_fc.head()

In [ ]:
df_fc.isnull().sum()

#common_neighbors            0
preferential attachment      0
resource_allocation_index    0
jaccard_coefficient          0
dtype: int64

In [ ]:
df_fc.shape

(488446, 4)

In [ ]:
df_eval = df_fc.join(future_connections, how='outer')

In [ ]:
df_eval.shape

(488446, 5)

In [ ]:
df_eval.head()

#common_neighbors  preferential attachment  \
(6, 840)                  9.0                     2070   
(4, 197)                  2.0                     3552   
(620, 979)                0.0                       28   
(519, 872)                2.0                      299   
(382, 423)                0.0                      205   

            resource_allocation_index  jaccard_coefficient  Future Connection  
(6, 840)                     0.136721             0.073770                0.0  
(4, 197)                     0.008437             0.015504                0.0  
(620, 979)                   0.000000             0.000000                0.0  
(519, 872)                   0.039726             0.060606                0.0  
(382, 423)                   0.000000             0.000000                0.0

In [73]:
df_train = df_eval[df_eval['Future Connection'].notnull()]
df_test = df_eval[df_eval['Future Connection'].isnull()]

features=['#common_neighbors', 'preferential attachment', 'resource_allocation_index','jaccard_coefficient']
label='Future Connection'

X_train, X_test, y_train, y_test = (train_test_split(df_train[features], 
                                                 df_train[label], random_state=23))

In [ ]:
RFC_fc = RandomForestClassifier(n_estimators=400, random_state=24).fit(X_train, y_train)

In [72]:
y_pred = RFC_fc.predict(X_test)
y_pred_prob = RFC_fc.predict_proba(X_test)

In [74]:
print('Train Accuracy_score: {}'.format(RFC_fc.score(X_train, y_train)))
print('Test Accuracy_score: {}'.format(RFC_fc.score(X_test, y_test)))
print('ROC_AUC_score: {}'.format(roc_auc_score(y_test,y_pred_prob[:,-1])))

Train Accuracy_score: 0.9913739763421292
Test Accuracy_score: 0.9560294374563243
ROC_AUC_score: 0.8930137334429934


In [80]:
y_pred_prob[:,1][:5]

array([ 0.26      ,  0.01      ,  0.01029392,  0.00083333,  0.00443766])

In [75]:
df_test.shape

(122112, 5)

In [76]:
df_test.head()

#common_neighbors  preferential attachment  \
(107, 348)                2.0                      884   
(542, 751)                0.0                      126   
(20, 426)                10.0                     4440   
(50, 989)                 0.0                       68   
(942, 986)                0.0                        6   

            resource_allocation_index  jaccard_coefficient  Future Connection  
(107, 348)                   0.025562             0.009009                NaN  
(542, 751)                   0.000000             0.000000                NaN  
(20, 426)                    0.082016             0.081967                NaN  
(50, 989)                    0.000000             0.000000                NaN  
(942, 986)                   0.000000             0.000000                NaN

In [77]:
df_test.tail()

#common_neighbors  preferential attachment  \
(165, 923)                0.0                      714   
(673, 755)                0.0                        3   
(939, 940)                0.0                        6   
(555, 905)                0.0                      160   
(75, 101)                 1.0                      638   

            resource_allocation_index  jaccard_coefficient  Future Connection  
(165, 923)                   0.000000             0.000000                NaN  
(673, 755)                   0.000000             0.000000                NaN  
(939, 940)                   0.000000             0.000000                NaN  
(555, 905)                   0.000000             0.000000                NaN  
(75, 101)                    0.008333             0.020833                NaN

Using network `G` and `future_connections`, identify the edges in `future_connections` with missing values and predict whether or not these edges will have a future connection.

To accomplish this, you will need to create a matrix of features for the edges found in `future_connections` using networkx, train a sklearn classifier on those edges in `future_connections` that have `Future Connection` data, and predict a probability of the edge being a future connection for those edges in `future_connections` where `Future Connection` is missing.



Your predictions will need to be given as the probability of the corresponding edge being a future connection.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.88 or higher will receive full points, and with an AUC of 0.82 or higher will pass (get 80% of the full points).

Using your trained classifier, return a series of length 122112 with the data being the probability of the edge being a future connection, and the index being the edge as represented by a tuple of nodes.

    Example:
    
        (107, 348)    0.35
        (542, 751)    0.40
        (20, 426)     0.55
        (50, 989)     0.35
                  ...
        (939, 940)    0.15
        (555, 905)    0.35
        (75, 101)     0.65
        Length: 122112, dtype: float64

In [82]:
def new_connections_predictions():
    
    # Your Code Here
    import sklearn
    from sklearn.linear_model import LogisticRegression
    from sklearn.ensemble import RandomForestClassifier
    from sklearn.metrics import roc_auc_score
    from sklearn.model_selection import train_test_split
    
    df_fc = pd.DataFrame(index=future_connections.index)
    for (u,v) in df_fc.index:
        df_fc.loc[(u,v),'#common_neighbors'] = len(list(nx.common_neighbors(G, u, v)))
    df_fc['preferential attachment'] = [i[2] for i in nx.preferential_attachment(G, df_fc.index)]
    df_fc['resource_allocation_index'] = [i[2] for i in nx.resource_allocation_index(G, df_fc.index)]
    df_fc['jaccard_coefficient'] = [i[2] for i in nx.jaccard_coefficient(G, df_fc.index)]
    df_eval = df_fc.join(future_connections, how='outer')
    df_train = df_eval[df_eval['Future Connection'].notnull()]
    df_test = df_eval[df_eval['Future Connection'].isnull()]

    features=['#common_neighbors', 'preferential attachment', 'resource_allocation_index','jaccard_coefficient']
    label='Future Connection'

    X_train, X_test, y_train, y_test = train_test_split(df_train[features], df_train[label], random_state=23)
    
    RFC_fc = RandomForestClassifier(n_estimators=400, random_state=24).fit(X_train, y_train)
    
    X_sub = df_test[features]

    df_test['Prob_0'] = RFC_fc.predict_proba(X_sub)[:,0]
    df_test['Prob_1'] = RFC_fc.predict_proba(X_sub)[:,1]
    df_test['Prob_1'] = df_test['Prob_1'].map(lambda x: round(x,2))
    
    return df_test['Prob_1'] # Your Answer Here